btx/processing/freqdir.py

```diff
@@ -0,0 +1,802 @@
+    over, and the algorithm will adjust the rank of the sketch to meet this error bound
+    as data streams in. The module also gives users the ability to perform the sketching
+    process over thresholded and non-zero image data.
+
+    [1] Frequent Directions: Simple and Deterministic Matrix 
+    Sketching Mina Ghashami, Edo Liberty, Jeff M. Phillips, and 
+    David P. Woodruff SIAM Journal on Computing 2016 45:5, 1762-1792
+
+    [2] Ghashami, M., Desai, A., Phillips, J.M. (2014). Improved 
+    Practical Matrix Sketching with Guarantees. In: Schulz, A.S., 
+    Wagner, D. (eds) Algorithms - ESA 2014. ESA 2014. Lecture Notes 
+    in Computer Science, vol 8737. Springer, Berlin, Heidelberg. 
+    https://doi.org/10.1007/978-3-662-44777-2_39
+
+    Attributes
+    ----------
+       start_offset: starting index of images to process
+       total_imgs: total number of images to process
+       ell: number of components of matrix sketch
+       alpha: proportion  of components to not rotate in frequent directions algorithm
+       exp, run, det_type: experiment properties
+       rankAdapt: indicates whether to perform rank adaptive FD
+       increaseEll: internal variable indicating whether ell should be increased for rank adaption
+       dir: directory to write output
+       merger: indicates whether object will be used to merge other FD objects
+       mergerFeatures: used if merger is true and indicates number of features of local matrix sketches
+       downsample, bin_factor: whether data should be downsampled and by how much
+       threshold: whether data should be thresholded (zero if less than threshold amount)
+       normalizeIntensity: whether data should be normalized to have total intensity of one
+       noZeroIntensity: whether data with low total intensity should be discarded
+       d: number of features (pixels) in data
+       m: internal frequent directions variable recording total number of components used in algorithm
+       sketch: numpy array housing current matrix sketch
+       mean: geometric mean of data processed
+       num_incorporated_images: number of images processed so far
+       imgsTracked: indices of images processed so far
+       currRun: Current datetime used to identify run
+    """
+
+    def __init__(
+        self,
+        start_offset,
+        total_imgs,
+        exp,
+        run,
+        det_type,
+        dir,
+        currRun,
+        ell=0, 
+        alpha=0,
+        rankAdapt=False,
+        merger=False,
+        mergerFeatures=0,
+        downsample=False,
+        bin_factor=2,
+        threshold=False,
+        normalizeIntensity=False,
+        noZeroIntensity=False, 
+    ):
+
+        self.comm = MPI.COMM_WORLD
+        self.rank = self.comm.Get_rank()
+        self.size = self.comm.Get_size()
+
+        self.currRun = currRun
+
+        self.merger = merger
+        if not merger:
```

Comment:
maybe directly do:
```python
if not self.merger:
```

Reply:
and even, maybe do `if self.merger:` first, it's easier to think about the logic